[PLUGIN] Github Deployments Plugin

anderoo opened the pull request:
## Hey, I just made a Pull Request!

Github Deployments Plugin which will display a table of the the last x deployments. At the moment I'm keeping this quite simple with the intention of adding custom columns and expose more data from the request later on.

Issue: #4794

<img width="778" alt="github-deployments-card" src="https://user-images.githubusercontent.com/57673895/112504521-c5946100-8d83-11eb-817a-475b0b748ca8.png">

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

adamdmharvey reviewed (commented):
This looks awesome; well done!

I recently had feedback to a PR saying I missed some composability work. In it, I was passing entity to my component, which you seem to also, and instead may need to use `useEntity()` inside of your component?  I'm 100% on it but just a note.

I also made multiple suggestions on case as GItHub refers to themselves as capital G and capital H so important to keep aligned, I think.

And a final point, is about the expectation of the slug.  This has been discussed elsewhere about actions as well, but this slug is used already to a) identify if an entity has code, b) if an entity uses actions, and potentially with this plugin now c) if an entity has deployments.  However, a team may only use one of those three, but the mere existence of this slug now if all the core feature and the two plugins (actions + deployments) exist may try to show data and error out even if there IS not intended to be any data.

Given that last point, I'd just ensure that this table prints hopefully not an error, per se, if deployments don't exist, but simply indicates there ARE no deployments for this entity?  that way if a team uses GitHub, and has say 100 entities, but only 10 use the deployments feature, the other 90 don't all start showing an error widget - if that makes sense?

All in all a great contrib to continue supporting more and more features of GitHub.  👏🏻 Well done! 🎉

freben commented:
Two practicalities - you need to sign your commits, see the info linked by the DCO build check - also, there's a small conflict in `packages/app/package.json`

anderoo commented on plugins/github-deployments/package.json:
```diff
@@ -0,0 +1,51 @@
+    "@material-ui/lab": "4.0.0-alpha.45",
+    "@octokit/graphql": "^4.6.1",
+    "moment": "^2.29.1",
+    "nock": "^13.0.11",
```
Yeah, I don't think this should even be here anymore. Good spot.

anderoo commented on plugins/github-deployments/src/components/GithubDeploymentsTable/GithubDeploymentsTable.tsx:
```diff
@@ -0,0 +1,90 @@
+  },
+  {
+    title: 'Status',
+    field: 'environment',
```
Good spot, the render makes it useless. Forgot to remove the field.

anderoo commented:
> This looks awesome; well done!
> 
> I recently had feedback to a PR saying I missed some composability work. In it, I was passing entity to my component, which you seem to also, and instead may need to use `useEntity()` inside of your component? I'm 100% on it but just a note.
> 
> I also made multiple suggestions on case as GItHub refers to themselves as capital G and capital H so important to keep aligned, I think.
> 
> And a final point, is about the expectation of the slug. This has been discussed elsewhere about actions as well, but this slug is used already to a) identify if an entity has code, b) if an entity uses actions, and potentially with this plugin now c) if an entity has deployments. However, a team may only use one of those three, but the mere existence of this slug now if all the core feature and the two plugins (actions + deployments) exist may try to show data and error out even if there IS not intended to be any data.
> 
> Given that last point, I'd just ensure that this table prints hopefully not an error, per se, if deployments don't exist, but simply indicates there ARE no deployments for this entity? that way if a team uses GitHub, and has say 100 entities, but only 10 use the deployments feature, the other 90 don't all start showing an error widget - if that makes sense?
> 
> All in all a great contrib to continue supporting more and more features of GitHub. 👏🏻 Well done! 🎉

Hey @adamdmharvey, many thanks for the helpful feedback. I've updated to use `useEntity` and fixed the incorrect names.

Completely agree on the project_slug issue. I think it's a trade off of whether we want to have many annotations which point to the same slug or have some entities which show empty state where annotations are not applicable. For now I've added an `<EmptyState>` when deployments are undefined or empty and added a test for this as well.

adamdmharvey commented on plugins/github-deployments/README.md:
```diff
@@ -0,0 +1,75 @@
+  <Grid container spacing={3} alignItems="stretch">
+    // ...
+    <Grid item xs={12} sm={6} md={4}>
+      <EntityGithubDeploymentsCard entity={entity} />
```
Think this can lose the entity pass now? or at least you can confirm with the code you have used to generate your screen shot.

anderoo replied:
Good spot. I'll update this and rerun local testing for screenshot.

adamdmharvey reviewed (approved):
Looks good to me! (one comment inline)

> Completely agree on the project_slug issue. I think it's a trade off of whether we want to have many annotations which point to the same slug or have some entities which show empty state where annotations are not applicable. For now I've added an <EmptyState> when deployments are undefined or empty and added a test for this as well.

There was discussion about maybe the GitHub Actions plugin does an extra call, like check the REST API to see if there are any available workflows, before truly "activating" the feature. Wonder if in future this could be refactored for GitHub Deployments in a similar way - check 1) if it has a slug, and 2) if deployments are enabled? If it doesn't work like that (as I'm not familiar with that part of the GitHub API), then maybe this is just easier than actions and simply saying there ARE no deployments (so long as it's not presented to the user as "an error") may be fine.

Anyway, looks great!!!

benjdlambert commented on plugins/github-deployments/src/api/index.ts:
```diff
@@ -0,0 +1,89 @@
+  }): Promise<GithubDeployment[]> {
+    const proxyUrl = await this.getProxyUrl();
+    const graphQlWithBaseURL = graphql.defaults({
+      baseUrl: `${proxyUrl}/github/api`,
```
@backstage/maintainers - I'm not a fan of the proxy use here with the PAT.

Should we be using the user token in the client to make this request and do it like the Github Actions and Pull Requests?

https://github.com/backstage/backstage/blob/master/plugins/github-actions/src/api/GithubActionsClient.ts#L28

anderoo replied:
Happy to change this to using the user token.

anderoo replied:
Updated for user token on client side.

anderoo replied:
However I'm also conscious that the restriction may be a blocker:
`There is a limit of 100 apps for one OAuth client/token pair`

anderoo commented on plugins/github-deployments/src/api/index.ts:
```diff
@@ -0,0 +1,99 @@
+    });
+
+    const response: any = await graphQLWithAuth(deploymentsQuery, options);
+    return response.repository?.deployments?.nodes?.reverse() || [];
```
Might also want to sort to ensure frequently deployed envs don't hide other envs. IN_PROGRESS > PENDING > ACTIVE > ... But this could be a follow up PR.

anderoo replied:
Yeah I'll do this in a follow up PR - there's a few other things I'd like to add: Refresh button, possible filter for status.

benjdlambert commented on plugins/github-deployments/src/api/index.ts:
```diff
@@ -0,0 +1,99 @@
+    last: number;
+  }): Promise<GithubDeployment[]> {
+    const token = await this.githubAuthApi.getAccessToken(['repo']);
+    const baseUrl = this.getBaseUrl() || 'https://api.github.com';
```
I think you should use `ScmIntegrations.fromConfig` here and pass in the `config`.

Then you can do `Integrations.byUrl(...).apiBaseUrl`.

This way if we change the integrations config, that class with still work :)

freben replied:
or `integrations.github.byUrl` if you want specifically only the github ones with the right types

anderoo replied:
I've removed baseUrl for now as we're not currently supporting GHE - See comment below.

anderoo replied:
So I've removed this for now as the current implementation doesn't support GHE or self hosted. I'll make a separate PR later on to add this.

benjdlambert commented on plugins/github-deployments/src/api/index.ts:
```diff
@@ -0,0 +1,99 @@
+      },
+    });
+
+    const response: any = await graphQLWithAuth(deploymentsQuery, options);
```
Do we need this `any` here? What's the types?

anderoo replied:
Updated for type.

benjdlambert commented on plugins/github-deployments/src/components/GithubDeploymentsCard.test.tsx:
```diff
@@ -0,0 +1,120 @@
+
+  describe('GithubDeploymentsCard', () => {
+    it('should display fetched data', async () => {
+      worker.use(rest.post('*', (_, res, ctx) => res(ctx.json(responseStub))));
```
`msw` has great support for GraphQL. I'd recommend using that as you can test much more than just a POST.

anderoo replied:
Thanks, added msw/graphql now.

benjdlambert commented on plugins/github-deployments/src/components/GithubDeploymentsTable/GithubDeploymentsTable.tsx:
```diff
@@ -0,0 +1,106 @@
+  },
+  {
+    title: 'Status',
+    render: (row: any): React.ReactNode => <State value={row.state} />,
```
Can we type this row instead of `any`?

anderoo replied:
I don't think so because of the type on render: `(property) Column<{}>.render?: ((data: {}, type: "row" | "group") => any) | undefined`

benjdlambert replied:
I think if you do `TableColumn<{ state: string}>[]` does it clear up the types and you can remove the any?

benjdlambert replied:
Here: https://github.com/backstage/backstage/pull/5114/files#diff-85ac04c660eadef50c087c0d75ba979a2f7411558ae3a1146664f0c23c2f5b41R56

anderoo replied:
Ah, nice. That works, I've updated it now.

benjdlambert commented on plugins/github-deployments/README.md:
```diff
@@ -0,0 +1,76 @@
+  description: |
+    Backstage is an open-source developer portal that puts the developer experience first.
+  annotations:
+    github.com/project-slug: YOUR_PROJECT_SLUG
```
How does this work for GHE and mainly self hosted Github? or is that not supported in this version?

anderoo replied:
I think for now I'd rather not support GHE and approach that in another PR - I've removed the baseUrl configuration for now.

anderoo commented:
@benjdlambert Thanks for the feedback. I've updated the code as per comments. For now I've removed the baseUrl access as I'll want to add support for GHE and self hosted in a future PR.

benjdlambert commented:
@freben do you want to have another quick look through?

benjdlambert commented on plugins/github-deployments/src/index.ts:
```diff
@@ -0,0 +1,20 @@
+ * limitations under the License.
+ */
+export {
+  githubDeploymentsPlugin as plugin,
```
Last thing from me - not sure we should really remove the specificity here, as we end up having to always rename them where we use them?

I think I like `githubDeploymentsPlugin` and `isGithubDeploymentsAvailable`. What do you think?

freben replied:
yes, this should definitely be keeping its original name, i started commenting in the app and docs code above to that effect as well but i'll just leave my agreeing in here instead

freben commented on plugins/github-deployments/README.md:
````diff
@@ -0,0 +1,67 @@
+```bash
+export AUTH_GITHUB_CLIENT_ID={{YOUR_CLIENT_ID}}
+export AUTH_CLIENT_SECRET={{YOUR_CLIENT_SECRET}}
+```
````
i think this section could be replaced with just a prerequisites link to https://backstage.io/docs/auth/github/provider ?

freben commented on plugins/github-deployments/README.md:
````diff
@@ -0,0 +1,67 @@
+```typescript
+// packages/app/src/plugins.ts
+
+export { plugin as GithubDeployments } from '@backstage/plugin-github-deployments';
````
but again, these packages should already have the correct name exported in the first place, ideally

freben commented on plugins/github-deployments/package.json:
```diff
@@ -0,0 +1,51 @@
+    "@material-ui/icons": "^4.9.1",
+    "@material-ui/lab": "4.0.0-alpha.45",
+    "@octokit/graphql": "^4.6.1",
+    "moment": "^2.29.1",
```
https://backstage.io/docs/architecture-decisions/adrs-adr010

we really should not be pulling in `moment`.

freben commented on plugins/github-deployments/src/components/GithubDeploymentsCard.test.tsx:
```diff
@@ -0,0 +1,123 @@
+  const worker = setupServer();
+  beforeAll(() => worker.listen());
+  afterAll(() => worker.close());
+  afterEach(() => worker.resetHandlers());
```
there's a `import { msw } from '@backstage/test-utils';` and then `msw.setupDefaultHandlers()` - seems like not a big deal but it does add for example stricter `onUnhandledRequest`, so might be worthwhile using that one instead of this triplet. See other places in the code for examples of its use

freben commented on plugins/github-deployments/src/components/GithubDeploymentsCard.test.tsx:
```diff
@@ -0,0 +1,123 @@
+      );
+
+      const rendered = render(
+        wrapInTestApp(
```
there's also a `renderInTestApp` :) that you have to await

using that one is actually preferable because it automatically wraps the rendering in such a way that side effects are properly caught with `act` etc

freben commented on plugins/github-deployments/src/components/GithubDeploymentsCard.tsx:
```diff
@@ -0,0 +1,73 @@
+  if (loading) {
+    return (
+      <InfoCard title="GitHub Deployments">
+        <LinearProgress />
```
use `Progress` from `core`, it's basically the same thing but makes sure that we can make all progress in the entire app behave the exact same way and look the same

freben commented on plugins/github-deployments/src/components/GithubDeploymentsCard.tsx:
```diff
@@ -0,0 +1,73 @@
+  if (error) {
+    return (
+      <InfoCard title="GitHub Deployments">
+        Error occured while fetching Data
```
no info at all about why? :) Maybe `<ResponseErrorPanel error={error} />` could be helpful (also from `core`)

freben commented on plugins/github-deployments/src/components/GithubDeploymentsTable/GithubDeploymentsTable.tsx:
```diff
@@ -0,0 +1,109 @@
+  const colorMap: Record<string, string> = {
+    PENDING: 'orange',
+    IN_PROGRESS: 'orange',
+    ACTIVE: 'green',
```
The `BackstageTheme` actually exposes `theme.palette.status.*` with a number of values.

But, even more interestingly, the concept of "status dots" is also available from core in the form of the `<StatusWarning />`, `<StatusPending />` etc components that you should use to get a consistent look-and-feel

anderoo replied:
👍 Nice - Updated.

freben commented on plugins/github-deployments/src/mocks/mocks.ts:
```diff
@@ -0,0 +1,67 @@
+import { QueryResponse } from '../api';
```
this should be below the copyright header :)

anderoo replied:
Auto import :(

anderoo commented:
@freben Thanks for the feedback. I've hopefully addressed all of the comments 👍

benjdlambert commented:
@freben for final approval? :)

anderoo commented:
@benjaminr Many thanks 👏

freben commented on yarn.lock:
```diff
@@ -4290,6 +4290,15 @@
     "@octokit/types" "^6.0.3"
     universal-user-agent "^6.0.0"
 
+"@octokit/graphql@^4.6.1":
```
so one final ask - could you update your dependency to be `^4.5.8` for this one and `yarn install` again? So that we don't have to get two versions of `@octokit/graphql`. And then later on, they will all be auto-bumped in unison by dependabot etc

anderoo replied:
Done 👍

anderoo replied:
Ok, should be good 👍

freben replied:
Nice. Now I wish I had a "merge automatically when it goes green unless further changes are made" button.

freben replied:
Oh wait, it just did, ok then :)

anderoo replied:
😆

freben reviewed (approved):
Just a final ask on the duplicate versions of octokit, then good to merge!
